causal_testing/testing/validation.py

```diff
@@ -0,0 +1,56 @@
+"""This module contains the CausalValidator class for performing Quantitive Bias Analysis techniques"""
+import math
+import numpy as np
+from scipy.stats import t
+from statsmodels.regression.linear_model import RegressionResultsWrapper
+
+
+class CausalValidator:
+    """A suite of validation tools to perform Quantitive Bias Analysis to back up causal claims"""
+
+    def estimate_robustness(self, model: RegressionResultsWrapper, q=1, alpha=1):
+        """Calculate the robustness of a linear regression model. This allow
+        the user to identify how large an unidentified confounding variable
+        would need to be to nullify the causal relationship under test."""
+
+        dof = model.df_resid
+        t_values = model.tvalues
+
+        fq = q * abs(t_values / math.sqrt(dof))
+        f_crit = abs(t.ppf(alpha / 2, dof - 1)) / math.sqrt(dof - 1)
+        fqa = fq - f_crit
+
+        rv = 0.5 * (np.sqrt(fqa**4 + (4 * fqa**2)) - fqa**2)
+
+        return rv
+
+    def estimate_e_value(
+        self, risk_ratio, confidence_intervals: tuple[float, float]
+    ) -> tuple[float, tuple[float, float]]:
+        """Calculate the E value from a risk ratio. This allow
+        the user to identify how large a risk an unidentified confounding
+        variable would need to be to nullify the causal relationship
+        under test."""
+
+        if risk_ratio >= 1:
+            e = risk_ratio + math.sqrt(risk_ratio * (risk_ratio - 1))
+
+            lower_limit = confidence_intervals[0]
+            if lower_limit <= 1:
```

Comment:
Should lower and upper limit both be testing for 1 in this way? Purely guessing from the variable names, I'd have said that we should have `lower_limit <= something <= upper_limit`, but with these tests, we'd end up with lower limit being greater than upper limit, or them both being set to 1.

Reply:
Mis implementation on my part. lower and upper limits shouldn't be returned or changed, they should only affect the e value if present. I have reimplemented this as both a with and without ci's method. Table 1 of https://oce.ovid.com/article/00000605-201708150-00007/PDF has been followed for the current implementation